src/Host/Monitor/Impl/CredentialManager.cs

```diff
@@ -0,0 +1,388 @@
+                        StringBuilder domain = new StringBuilder(CRED_MAX_USERNAME_LENGTH);
+                        StringBuilder password = new StringBuilder(CRED_MAX_USERNAME_LENGTH);
+
+                        int usernameLen = username.Capacity;
+                        int domainLen = domain.Capacity;
+                        int passwordLen = password.Capacity;
+
+                        if (!CredUnPackAuthenticationBuffer(
+                            0,
+                            credentialBuffer,
+                            credentialSize,
+                            username,
+                            ref usernameLen,
+                            domain,
+                            ref domainLen,
+                            password,
+                            ref passwordLen)) {
+                            throw new Win32Exception(Marshal.GetLastWin32Error());
+                        }
+
+                        IntPtr token = IntPtr.Zero;
+                        try {
+                            var usernameBldr = new StringBuilder(257);
+                            var domainBldr = new StringBuilder(257);
+
+                            uint error = CredUIParseUserName(username.ToString(), usernameBldr, usernameBldr.Capacity, domainBldr, domainBldr.Capacity);
+                            if (error != 0) {
+                                throw new Win32Exception((int)error);
+                            }
+
+                            if (LogonUser(usernameBldr.ToString(), domainBldr.ToString(), password.ToString(), (int)LogonType.LOGON32_LOGON_INTERACTIVE, (int)LogonProvider.LOGON32_PROVIDER_DEFAULT, ref token)) {
+                                Credential creds = new Credential();
+                                creds.targetName = BrokerUserCredName;
+                                creds.type = (int)CredType.GENERIC;
+                                creds.userName = username.ToString();
+                                creds.attributeCount = 0;
+                                creds.persist = (int)CredPersist.LOCAL_MACHINE;
+                                byte[] bpassword = Encoding.Unicode.GetBytes(password.ToString());
+                                creds.credentialBlobSize = bpassword.Length;
+                                creds.credentialBlob = Marshal.StringToCoTaskMemUni(password.ToString());
+
+                                if (!CredWrite(ref creds, 0)) {
+                                    throw new Win32Exception(Marshal.GetLastWin32Error());
+                                }
+                            }
+                        } finally {
+                            if (token != IntPtr.Zero) {
+                                CloseHandle(token);
+                            }
+                        }
+
+                        break;
+                    }
+
+                    // CredUIPromptForWindowsCredentials failed to load some component
+                    if (ret != 0) {
```

Comment:
Should this handle the return error codes for all the other `Cred*` functions here? E.g. if `CredUIParseUserName`, we just throw with the error code provided, without a custom error message - is that right?